presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationCompiler.java

```diff
@@ -254,4 +230,209 @@ private static Method getCombineFunction(Class<?> clazz, Class<?> stateClass)
 
         return stateClasses;
     }
+
+    private static String getDescription(AnnotatedElement base, AnnotatedElement override)
+    {
+        Description description = override.getAnnotation(Description.class);
+        if (description != null) {
+            return description.value();
+        }
+        description = base.getAnnotation(Description.class);
+        return (description == null) ? null : description.value();
+    }
+
+    private static Set<String> getLiteralParameter(Method inputFunction)
+    {
+        ImmutableSet.Builder<String> literalParametersBuilder = ImmutableSet.builder();
+
+        Annotation[] literalParameters = inputFunction.getAnnotations();
+        for (Annotation annotation : literalParameters) {
+            if (annotation instanceof LiteralParameters) {
+                for (String literal : ((LiteralParameters) annotation).value()) {
+                   literalParametersBuilder.add(literal);
+                }
+            }
+        }
+
+        return literalParametersBuilder.build();
+    }
+
+    private static List<Method> findPublicStaticMethodsWithAnnotation(Class<?> clazz, Class<?> annotationClass)
+    {
+        ImmutableList.Builder<Method> methods = ImmutableList.builder();
+        for (Method method : clazz.getMethods()) {
+            for (Annotation annotation : method.getAnnotations()) {
+                if (annotationClass.isInstance(annotation)) {
+                    checkArgument(Modifier.isStatic(method.getModifiers()) && Modifier.isPublic(method.getModifiers()), "%s annotated with %s must be static and public", method.getName(), annotationClass.getSimpleName());
+                    methods.add(method);
+                }
+            }
+        }
+        return methods.build();
+    }
+
+    public static class BindableAggregationFunction
```

Comment:
Looking at this more carefully, why do we need this class? It seems to only be used by `AnnotationBasedSqlAggregationFunction` which is now just a wrapper around `BindableAggregationFunction`. I didn't read through all the details, but it seems like it would be better to have `AnnotationBasedSqlAggregationFunction` call a method on `AggregationCompiler` to get a `AggregationMetadata` and then it could modify that structure and call the compiler again in `specialize()`